src/loaders/zarr_utils/ChunkPrefetchIterator.ts

```diff
@@ -0,0 +1,119 @@
+  end: number;
+};
+
+const skipC = (idx: number): number => idx + Number(idx !== 0);
+const directionToIndex = (dir: PrefetchDirection): number => skipC(dir >> 1);
+
+/**
+ * Since the user is most likely to want nearby data (in space or time) first, we should prefetch those chunks first.
+ *
+ * Given a list of just-loaded chunks and some bounds, `ChunkPrefetchIterator` iterates evenly outwards in T/Z/Y/X.
+ */
+// NOTE: Assumes `chunks` form a rectangular prism! Will create gaps otherwise! (in practice they always should)
+export default class ChunkPrefetchIterator {
+  directions: PrefetchDirectionState[];
+
+  constructor(chunks: TCZYX<number>[], tzyxMaxOffset: TZYX, tzyxChunkSize: TZYX) {
+    // Get max and min chunk coordinates for T/Z/Y/X
+    const extrema = [Infinity, -Infinity, Infinity, -Infinity, Infinity, -Infinity, Infinity, -Infinity];
+
+    for (const chunk of chunks) {
+      for (let j = 0; j < 4; j++) {
+        const val = chunk[skipC(j)];
+        const extremaIdx = j << 1;
+
+        if (val < extrema[extremaIdx]) {
+          extrema[extremaIdx] = val;
+        }
+
+        if (val > extrema[extremaIdx + 1]) {
+          extrema[extremaIdx + 1] = val;
+        }
+      }
+    }
+
+    // Create `PrefetchDirectionState`s for each direction
+    const directions: PrefetchDirectionState[] = extrema.map((start, direction) => {
+      const dimension = direction >> 1;
+      if (direction & 1) {
+        // Positive direction - end is either the max coordinate in the fetched set plus the max offset in this
+        // dimension, or the max chunk coordinate in this dimension, whichever comes first
+        const end = Math.min(start + tzyxMaxOffset[dimension], tzyxChunkSize[dimension] - 1);
+        return { direction, start, end, chunks: [] };
+      } else {
+        // Negative direction - end is either the min coordinate in the fetched set minus the max offset in this
+        // dimension, or 0, whichever comes first
+        const end = Math.max(start - tzyxMaxOffset[dimension], 0);
+        return { direction, start, end, chunks: [] };
+      }
+    });
+
+    // Fill each `PrefetchDirectionState` with chunks at the border of the fetched set
+    for (const chunk of chunks) {
+      for (const dir of directions) {
+        if (chunk[directionToIndex(dir.direction)] === dir.start) {
+          dir.chunks.push(chunk);
+        }
+      }
+    }
+
+    this.directions = directions;
+  }
+
+  *[Symbol.iterator](): Iterator<TCZYX<number>> {
+    let offset = 1;
+
+    while (this.directions.length > 0) {
+      // Remove directions in which we have hit a boundary
+      this.directions = this.directions.filter((dir) => {
+        if (dir.direction & 1) {
+          return dir.start + offset <= dir.end;
+        } else {
+          return dir.start - offset >= dir.end;
+        }
+      });
+
+      // Yield chunks one chunk farther out in every remaining direction
+      for (const dir of this.directions) {
```

Comment:
to make sure I understand, is there currently an implied preferential order in T,Z,Y,X ?

Reply:
Yes